File: typescript/test/payer_proof.test.ts
/**
 * The channel-open payer proof, bound to the shared protocol vector.
 *
 * The hub rebuilds this message independently and compares recovered signers, so "close
 * enough" is indistinguishable from wrong: the vector is the contract, and these tests
 * assert against its bytes rather than against our own implementation.
 */

import { describe, expect, it } from 'vitest';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { privateKeyToAccount } from 'viem/accounts';
import type { Hex } from 'viem';

import {
  canonicalProofAmountCents,
  channelOpenProofMessage,
  recoverChannelOpenPayer,
  signChannelOpenProof,
} from '../src/payerProof';

interface VectorCase {
  name: string;
  chain: string;
  tx_hash: string;
  payer: string;
  amount_usd: number;
  amount_cents: number;
  message: string;
  signature?: Hex;
  recovers_to?: string;
}

const vector = JSON.parse(
  readFileSync(
    join(__dirname, '../../../aimarket-protocol/test-vectors/payer-proof.json'),
    'utf8',
  ),
) as { test_key: Hex; test_address: string; cases: VectorCase[] };

describe('channel-open payer proof', () => {
  it('has cases to check', () => {
    expect(vector.cases.length).toBeGreaterThan(4);
  });

  for (const c of vector.cases) {
    it(`reproduces the canonical message: ${c.name}`, () => {
      expect(
        channelOpenProofMessage({
          chain: c.chain,
          txHash: c.tx_hash,
          payer: c.payer,
          amountUsd: c.amount_usd,
        }),
      ).toBe(c.message);
    });

    it(`converts the amount the way the hub does: ${c.name}`, () => {
      expect(canonicalProofAmountCents(c.amount_usd)).toBe(c.amount_cents);
    });
  }

  it('rounds a half-cent to EVEN, not away from zero', () => {
    // Math.round(12.5) is 13; the hub's Python round() gives 12. Using the wrong rule
    // produces a valid-looking signature the hub refuses.
    expect(canonicalProofAmountCents(0.125)).toBe(12);
    expect(canonicalProofAmountCents(0.135)).toBe(14);
    expect(Math.round(0.125 * 100)).toBe(13); // the trap, spelled out
  });

  it('refuses an unusable amount deterministically', () => {
    expect(canonicalProofAmountCents(Number.NaN)).toBe(-1);
    expect(canonicalProofAmountCents(Number.POSITIVE_INFINITY)).toBe(-1);
  });

  it('signs and recovers the vector signature', async () => {
    const account = privateKeyToAccount(vector.test_key);
    for (const c of vector.cases.filter((x) => x.signature)) {
      const params = {
        chain: c.chain,
        txHash: c.tx_hash,
        payer: c.payer,
        amountUsd: c.amount_usd,
      };
      const signature = await signChannelOpenProof(params, (message) =>
        account.signMessage({ message }),
      );
      expect(signature).toBe(c.signature);
      expect((await recoverChannelOpenPayer(params, signature)).toLowerCase()).toBe(
        c.recovers_to!.toLowerCase(),
      );
    }
  });

  it('a different wallet does not recover as the payer', async () => {
    const other = privateKeyToAccount(`0x${'77'.repeat(32)}` as Hex);
    const c = vector.cases[0];
    const params = {
      chain: c.chain,
      txHash: c.tx_hash,
      payer: c.payer,
      amountUsd: c.amount_usd,
    };
    const signature = await signChannelOpenProof(params, (m) => other.signMessage({ message: m }));
    const recovered = await recoverChannelOpenPayer(params, signature);
    expect(recovered.toLowerCase()).not.toBe(vector.test_address.toLowerCase());
  });
});

File: dart/test/payer_proof_test.dart
/// The channel-open payer proof, bound to the shared protocol vector.
///
/// The hub rebuilds this message independently and compares recovered signers, so the
/// vector is the contract: these tests assert against its bytes rather than against our
/// own implementation agreeing with itself.
library;

import 'dart:convert';
import 'dart:io';

import 'package:aimarket_agent/aimarket_agent.dart';
import 'package:test/test.dart';
import 'package:web3dart/credentials.dart';
import 'package:web3dart/crypto.dart';

void main() {
  final file = File('../../aimarket-protocol/test-vectors/payer-proof.json');
  if (!file.existsSync()) {
    // Skipped loudly rather than silently passing: without the vector these tests
    // would only prove the implementation agrees with itself.
    test('shared payer-proof vector is present', () {
      fail('missing ${file.path} — the cross-SDK contract cannot be checked');
    });
    return;
  }
  final vector = jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
  final cases = (vector['cases'] as List).cast<Map<String, dynamic>>();

  group('channel-open payer proof', () {
    test('the vector has cases to check', () {
      expect(cases.length, greaterThan(4));
    });

    for (final c in cases) {
      test('reproduces the canonical message: ${c['name']}', () {
        expect(
          channelOpenProofMessage(
            chain: c['chain'] as String,
            txHash: c['tx_hash'] as String,
            payer: c['payer'] as String,
            amountUsd: c['amount_usd'] as num,
          ),
          equals(c['message'] as String),
        );
      });

      test('converts the amount the way the hub does: ${c['name']}', () {
        expect(
          canonicalProofAmountCents(c['amount_usd'] as num),
          equals(c['amount_cents'] as int),
        );
      });
    }

    test('rounds a half-cent to EVEN, not away from zero', () {
      // (0.125 * 100).round() is 13 in Dart; the hub's Python round() gives 12, so the
      // naive rule produces a valid-looking signature the hub refuses.
      expect(canonicalProofAmountCents(0.125), equals(12));
      expect(canonicalProofAmountCents(0.135), equals(14));
      expect((0.125 * 100).round(), equals(13)); // the trap, spelled out
    });

    test('refuses an unusable amount deterministically', () {
      expect(canonicalProofAmountCents(double.nan), equals(-1));
      expect(canonicalProofAmountCents(double.infinity), equals(-1));
    });

    test('signs exactly what the vector signed', () {
      final key = vector['test_key'] as String;
      final signed = cases.where((x) => x.containsKey('signature')).toList();
      expect(signed, isNotEmpty);
      for (final c in signed) {
        expect(
          signChannelOpenProof(
            privateKeyHex: key,
            chain: c['chain'] as String,
            txHash: c['tx_hash'] as String,
            payer: c['payer'] as String,
            amountUsd: c['amount_usd'] as num,
          ),
          equals(c['signature'] as String),
          reason: 'the signature must match the vector byte for byte',
        );
      }
    });

    test('the signature recovers to the paying wallet', () {
      final key = vector['test_key'] as String;
      final c = cases.firstWhere((x) => x.containsKey('signature'));
      final params = (
        chain: c['chain'] as String,
        txHash: c['tx_hash'] as String,
        payer: c['payer'] as String,
        amountUsd: c['amount_usd'] as num,
      );
      final digest = channelOpenProofHash(
        chain: params.chain,
        txHash: params.txHash,
        payer: params.payer,
        amountUsd: params.amountUsd,
      );
      final sig = signChannelOpenProof(
        privateKeyHex: key,
        chain: params.chain,
        txHash: params.txHash,
        payer: params.payer,
        amountUsd: params.amountUsd,
      );
      final bytes = hexToBytes(sig.substring(2));
      final recovered = EthereumAddress.fromPublicKey(
        ecRecover(
          digest,
          MsgSignature(
            bytesToUnsignedInt(bytes.sublist(0, 32)),
            bytesToUnsignedInt(bytes.sublist(32, 64)),
            bytes[64],
          ),
        ),
      ).hexEip55.toLowerCase();
      expect(recovered, equals(EthPrivateKey.fromHex(key).address.hexEip55.toLowerCase()));
    });
  });
}
